### DISCUSSION TITLE
Is there a way to use $emit in boot function defintion?

### DISCUSSION THREAD

**Discussion by luzdealba** [quasarframework/quasar#14343] [Quasar App Extensions]
This is related to the [$emit on app extension](https://forum.quasar-framework.org/topic/7762/emit-on-app-extension/4) forum post.

Essentially, is there a way to invoke the `emit()` function from within a [boot file](https://quasar.dev/quasar-cli-vite/boot-files)?

**Comment by TobyMosque** [comment]
https://v3-migration.vuejs.org/breaking-changes/events-api.html#_3-x-update
but u can use a third party event library like mitt:

first of all, u can write a dumb inject helper to use in your boots:
```ts
import { App, InjectionKey } from 'vue';
export function inject<T>(app: App, key: InjectionKey<T>): T {
  return app._context.provides[key as never] as T;
}
```

so, create a boot who register the emitter, this boot would be called before the boots who requires the emitter.
```ts
import { boot } from 'quasar/wrappers';
import { InjectionKey } from 'vue';
import mitt, { Emitter } from 'mitt';

type Events = {
  foo: string;
  bar?: number;
};

export const emitterKey: InjectionKey<Emitter<Events>> = Symbol('emitter-key');
export default boot(({ app }) => {
  const emitter: Emitter<Events> = mitt<Events>();
  app.provide(emitterKey, emitter);
});
```

in other boots:
```ts
import { emitterKey } from './emitter'
import { boot } from 'quasar/wrappers';
import { inject } from 'src/utils';

export default boot(({ app }) => {
  const emitter = inject(app, emitterKey)
  emitter.on('foo', function (val: string) {
    console.log(val)
  })
  emitter.on('bar', function (val: number | undefined) {
    console.log(val)
  })
});
```

so, in your components:
```ts
import { emitterKey } from 'src/boot/emitter'
import { defineComponent, inject } from 'vue';

export default defineComponent({
  setup () {
    const emitter = inject(emitterKey)
    emitter.on('foo', function (val: string) {
      console.log(val)
    })
    emitter.on('bar', function (val: number | undefined) {
      console.log(val)
    })
  }
})
```

**Reply by luzdealba** [reply]
@TobyMosque I never thanked you for your detailed answer  🙏

For the record, I ended up having a helper function that would get the `emit` function as a parameter, and then invoking the emitter from inside the helper (that is, in whichever context the helper was used).
